core/src/main/java/org/infinispan/atomic/impl/AtomicKeySetImpl.java

```diff
@@ -554,4 +556,54 @@ public Externalizable readObject(ObjectInput input) throws IOException, ClassNot
          }
       }
    }
+
+   public static class Marshaller implements MessageMarshaller<AtomicKeySetImpl> {
+
+      private final GlobalComponentRegistry gcr;
+
+      public Marshaller(GlobalComponentRegistry gcr) {
+         this.gcr = gcr;
+      }
+
+      @Override
+      public AtomicKeySetImpl readFrom(ProtoStreamReader reader) throws IOException {
+         ByteString cacheName1 = ByteString.fromString(reader.readString("cachename"));
+         Object group = reader.readObject("group", UserObject.class).get();
+         ComponentRegistry cr = gcr.getNamedComponentRegistry(cacheName1);
+         InvocationContextFactory icf = cr.getComponent(InvocationContextFactory.class);
+         InvocationContext ctx = icf.createNonTxInvocationContext();
+         AsyncInterceptorChain chain = cr.getComponent(AsyncInterceptorChain.class);
+         GetKeysInGroupCommand cmd = cr.getCommandsFactory().buildGetKeysInGroupCommand(0, group);
+         Map<Object, Object> map = (Map<Object, Object>) chain.invoke(ctx, cmd);
+         // GetKeyInGroupCommand will set skip lookup for all entries in context, including the entry which
+         // is currently being loaded. We have to revert the setting or it won't be loaded.
+         CacheEntry entry = ctx.lookupEntry(group);
+         if (entry != null) {
+            entry.setSkipLookup(false);
+         }
+         // In a similar fashion the GetKeyInGroupCommand recorded version seen before it was loaded
+         // it recorded it as a non-existent version.
+         // The versions read map is not immutable, though, so we can undo that operation.
+         if (ctx.isInTxScope()) {
+            ((TxInvocationContext) ctx).getCacheTransaction().getVersionsRead().remove(group);
+         }
+         return new AtomicKeySetImpl(cacheName1, group, map.keySet(), null, null);
+      }
+
+      @Override
+      public void writeTo(ProtoStreamWriter writer, AtomicKeySetImpl atomicKeySet) throws IOException {
+         writer.writeString("cachename", atomicKeySet.cacheName.toString());
+         writer.writeObject("group", new UserObject(atomicKeySet.group), UserObject.class);
+      }
+
+      @Override
+      public Class<? extends AtomicKeySetImpl> getJavaClass() {
```

Comment:
getJavaClass() should declare the exact class. no need for wildcards. not just here, everywhere.

Reply:
Good point. This was just pulled in by intellij's "implement methods" as `BaseMarshaller::getJavaClass` is `Class<? extends T> getJavaClass();`.

I can update though.